src/jitlayers.cpp

```diff
@@ -165,12 +167,17 @@ void addOptimizationPasses(PassManager *PM)
 
     PM->add(createEarlyCSEPass()); //// ****
 
+    PM->add(createLoopIdiomPass()); //// ****
+    PM->add(createLoopRotatePass());           // Rotate loops.
 #ifdef USE_POLLY
+    // LCSSA (which has already run at this point due to the dependencies of the
+    // above passes) introduces redundant phis that hinder Polly. Therefore we
+    // run InstCombine here to remove them.
+    PM->add(createInstructionCombiningPass());
+    PM->add(polly::createCodePreparationPass());
     polly::registerPollyPasses(*PM);
+    PM->add(polly::createCodegenCleanupPass());
```

Comment:
Codegen cleanup and codegen prepare passes are partially independent changes. I would suggest to add them in separate commits (may be the still pull-request though).